### PR TITLE
Split frontend type system into frontend dependent and common part

### PR DESCRIPTION
**Pull request by petiaccja**

### CODE REVIEW

File: src/functional/type_system/type_info.py [hunk 1]
```diff
@@ -0,0 +1,584 @@
+import functools
```
Comment (tehrengruber) [inline src/functional/type_system/type_info.py:1]: Copyright header.

File: src/functional/type_system/type_specifications.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+from dataclasses import dataclass
```
Comment (tehrengruber) [inline src/functional/type_system/type_specifications.py:1]: Copyright header.

File: src/functional/ffront/type_specifications.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from dataclasses import dataclass
```
Comment (tehrengruber) [inline src/functional/ffront/type_specifications.py:1]: Copyright header.

File: tests/functional_tests/ffront_tests/test_foast_lowering.py [hunk 4]
```diff
@@ -15,8 +15,9 @@
 
 from types import SimpleNamespace
 
+import functional.ffront.type_specifications
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_foast_lowering.py:18]:
```suggestion
```
You already have the `ts` alias below.

File: tests/functional_tests/ffront_tests/test_foast_lowering.py [hunk 5]
```diff
@@ -221,10 +223,10 @@ def test_call():
     # create something that appears to the lowering like a field operator.
     #  we could also create an actual field operator, but we want to avoid
     #  using such heavy constructs for testing the lowering.
-    field_type = make_symbol_type_from_typing(Field[..., float64])
+    field_type = from_type_hint(Field[..., float64])
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_foast_lowering.py:226]:
```suggestion
    field_type = type_translation.from_type_hint(Field[..., float64])
```
For consistency.

File: tests/functional_tests/ffront_tests/test_foast_lowering.py [hunk 6]
```diff
@@ -28,7 +29,8 @@
 )
 from functional.ffront.foast_to_itir import FieldOperatorLowering
 from functional.ffront.func_to_foast import FieldOperatorParser
-from functional.ffront.symbol_makers import make_symbol_type_from_typing
+from functional.type_system import type_specifications as ts
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_foast_lowering.py:32]:
```suggestion
from functional.ffront.type_system import type_specifications as ts
```

File: tests/functional_tests/ffront_tests/test_foast_lowering.py [hunk 7]
```diff
@@ -221,10 +223,10 @@ def test_call():
     # create something that appears to the lowering like a field operator.
     #  we could also create an actual field operator, but we want to avoid
     #  using such heavy constructs for testing the lowering.
-    field_type = make_symbol_type_from_typing(Field[..., float64])
+    field_type = from_type_hint(Field[..., float64])
     identity = SimpleNamespace(
-        __gt_type__=lambda: ct.FieldOperatorType(
-            definition=ct.FunctionType(args=[field_type], kwargs={}, returns=field_type)
+        __gt_type__=lambda: functional.ffront.type_specifications.FieldOperatorType(
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_foast_lowering.py:228]:
```suggestion
        __gt_type__=lambda: ts.FieldOperatorType(
```

File: tests/functional_tests/ffront_tests/test_foast_parsing.py [hunk 8]
```diff
@@ -51,6 +50,8 @@
     tuple_get,
     xor_,
 )
+from functional.type_system import type_specifications as ts
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_foast_parsing.py:53]:
```suggestion
from functional.ffront.type_system import type_specifications as ts
```

File: tests/functional_tests/ffront_tests/test_past_parsing.py [hunk 9]
```diff
@@ -12,18 +12,18 @@
...
 from eve.pattern_matching import ObjectPattern as P
 from functional.common import Field, GTTypeError
-from functional.ffront import common_types, program_ast as past
+from functional.ffront import program_ast as past
 from functional.ffront.decorator import field_operator
-from functional.ffront.fbuiltins import Dimension, float64
+from functional.ffront.fbuiltins import float64
 from functional.ffront.func_to_past import ProgramParser
 from functional.ffront.past_passes.type_deduction import ProgramTypeError
+from functional.type_system import type_specifications as ts
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_past_parsing.py:26]:
```suggestion
from functional.ffront.type_system import type_specifications as ts
```

File: tests/functional_tests/ffront_tests/test_math_builtin_execution.py [hunk 10]
```diff
@@ -4,17 +4,14 @@
...
-)
+import functional.ffront.dialect_ast_enums as ast_enums
+from functional.ffront import fbuiltins, field_operator_ast as foast
 from functional.ffront.decorator import FieldOperator
-from functional.ffront.fbuiltins import Dimension, Field, float64, int32, int64
+from functional.ffront.fbuiltins import Dimension, Field, float64
 from functional.ffront.foast_passes.type_deduction import FieldOperatorTypeDeduction
 from functional.iterator.embedded import np_as_located_field
 from functional.program_processors.runners import roundtrip
+from functional.type_system import type_specifications as ts, type_translation
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_math_builtin_execution.py:14]:
```suggestion
from functional.ffront.type_system import type_specifications as ts
from functional.type_system import type_translation
```

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: setup.cfg [hunk 11]
```diff
@@ -236,7 +236,13 @@ ignore_errors = True
...
 
-[mypy-functional.ffront.symbol_makers]
+[mypy-functional.type_system.type_translation]
+ignore_errors = True
+
+[mypy-functional.type_system.type_info]
+ignore_errors = True
+
+[mypy-functional.ffront.foast_passes.type_deduction]
 ignore_errors = True
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply setup.cfg:246]:
```
src/functional/ffront/foast_passes/type_deduction.py:106:9: error: Unsupported right operand type for in ("Union[List[Dimension], ellipsis]")  [operator]
src/functional/ffront/foast_passes/type_deduction.py:106:45: error: Item "ellipsis" of "Union[List[Dimension], ellipsis]" has no attribute "__iter__" (not iterable)  [union-attr]
src/functional/ffront/foast_passes/type_deduction.py:177:83: error: Argument "returns" to "FunctionType" has incompatible type "TypeSpec"; expected "Union[DataType, DeferredType, VoidType]"  [arg-type]
src/functional/ffront/foast_passes/type_deduction.py:642:12: error: Unsupported right operand type for in ("Union[List[Dimension], ellipsis]")  [operator]
src/functional/ffront/foast_passes/type_deduction.py:643:60: error: Item "ellipsis" of "Union[List[Dimension], ellipsis]" has no attribute "__iter__" (not iterable)  [union-attr]
src/functional/ffront/foast_passes/type_deduction.py:651:34: error: Item "ellipsis" of "Union[List[Dimension], ellipsis]" has no attribute "__iter__" (not iterable)  [union-attr]
```

File: src/functional/ffront/dialect_ast_enums.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
+from eve import StrEnum
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply src/functional/ffront/dialect_ast_enums.py:15]: Done

File: src/functional/ffront/past_passes/closure_var_type_deduction.py [hunk 13]
```diff
@@ -42,7 +42,7 @@ def visit_Program(self, node: past.Program, **kwargs) -> past.Program:
                 new_symbol: past.Symbol = past.Symbol(
                     id=sym.id,
                     location=sym.location,
-                    type=make_symbol_type_from_value(self.closure_vars[sym.id]),
+                    type=from_value(self.closure_vars[sym.id]),
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply src/functional/ffront/past_passes/closure_var_type_deduction.py:45]: Done

File: src/functional/ffront/type_info.py [hunk 14]
```diff
@@ -1,321 +1,70 @@
...
+    "is_integral",
+    "is_logical",
+    "primitive_constituents",
+    "promote",
+    "promote_dims",
+    "return_type",
+    "return_type_field",
+    "return_type_func",
+    "type_class",
+]
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply src/functional/ffront/type_info.py:69]: Flake8 complains about unused imports.

File: src/functional/ffront/type_translation.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
+from functional.type_system.type_translation import from_type_hint, from_value, get_scalar_kind
+
+
+__all__ = ["get_scalar_kind", "from_type_hint", "from_value"]
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply src/functional/ffront/type_translation.py:18]: Not at the moment, so in essence, I can remove this. I have some ideas to replace `__gt_type__` with a non-intrusive but still extensible alternative, and that would eventually relocate the handling of field view types to this file.

File: src/functional/ffront/decorator.py [hunk 16]
```diff
@@ -31,15 +31,16 @@
 import numpy as np
 from devtools import debug
 
+import functional.ffront.dialect_ast_enums as ast_enums
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply src/functional/ffront/decorator.py]: Done

**Review by tehrengruber — APPROVED** [review]

**Pull request merged**